File: src/main/java/io/github/saneea/RootFeatureProvider.java
package io.github.saneea;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.lang.reflect.InvocationTargetException;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Properties;
import java.util.Set;

import io.github.saneea.textfunction.Utils;

public class RootFeatureProvider implements FeatureProvider {
	private final Map<String, String> featureAliases;

	public RootFeatureProvider() throws IOException {
		this.featureAliases = Utils.toMap(loadProperties());
	}

	private static Properties loadProperties() throws IOException {
		Properties featureAliases = new Properties();
		InputStream resourceAsStream = App.class.getResourceAsStream("/feature-aliases.properties");
		try (Reader reader = new InputStreamReader(//
				new BufferedInputStream(//
						resourceAsStream), //
				StandardCharsets.UTF_8)) {
			featureAliases.load(reader);
		}
		return featureAliases;
	}

	@Override
	public Set<String> featuresNames() {
		return featureAliases.keySet();
	}

	@Override
	public Feature createFeature(String featureName) throws ClassNotFoundException, InstantiationException,
			IllegalAccessException, InvocationTargetException, NoSuchMethodException {
		Class<?> featureClass = getFeatureClass(featureName);
		return featureClass == null//
				? null //
				: createFeatureInstance(featureClass);
	}

	private Class<?> getFeatureClass(String featureName) throws ClassNotFoundException {
		String featureClassName = featureAliases.get(featureName);
		return featureClassName == null//
				? null //
				: Class.forName(featureClassName);
	}

	private Feature createFeatureInstance(Class<?> featureClass)
			throws InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException {
		Object featureObj = featureClass.getDeclaredConstructor().newInstance();

		if (!(featureObj instanceof Feature)) {
			throw new IllegalArgumentException(
					"Class " + featureObj.getClass() + " does not implement " + Feature.class);
		}

		return (Feature) featureObj;
	}
}

File: src/main/kotlin/io/github/saneea/dvh/RootFeatureProvider.kt
package io.github.saneea.dvh

import io.github.saneea.dvh.utils.Utils
import java.nio.charset.StandardCharsets
import java.util.*

const val FEATURE_ALIASES_PROPS_FILE_NAME = "/feature-aliases.properties"

class RootFeatureProvider : FeatureProvider {

    private val featureAliases: Map<String, String> = Utils.toMap(loadProperties())

    override fun featuresNames() = featureAliases.keys

    override fun createFeature(featureName: String) =
        featureAliases[featureName]
            ?.let { createFeatureInstance(Class.forName(it)) }

    private fun createFeatureInstance(featureClass: Class<*>): Feature {
        val featureObj = featureClass.getDeclaredConstructor().newInstance()
        require(featureObj is Feature) {
            "Class ${featureObj.javaClass} does not implement ${Feature::class.java}"
        }
        return featureObj
    }

    companion object {

        private fun loadProperties(): Properties {
            val featureAliases = Properties()

            val resourceAsStream = App::class.java.getResourceAsStream(FEATURE_ALIASES_PROPS_FILE_NAME)

            requireNotNull(resourceAsStream) {
                "Feature aliases file is not found: $FEATURE_ALIASES_PROPS_FILE_NAME"
            }
                .buffered()
                .reader(StandardCharsets.UTF_8)
                .use(featureAliases::load)

            return featureAliases
        }
    }

}